### PR TITLE
Fix: LogicalExpression checks for short circuit (fixes #5693)

### LINKED ISSUE #5693: `no-constant-condition` should respect partial conditions too
Not sure if I got the title correctly, but we are valuating for a whole condition but if its part of multiple conditions, we are not checking them. 

I was working on #5228 and came across this bug . Please suggest if its a valid bug and I can take care fixing it. 

**What version of ESLint are you using?**
1.5.1
**What parser (default, Babel-ESLint, etc.) are you using?**
default
**Please show your full configuration:**
```js
{
  "rules": {
    "no-constant-condition": 2   // disallow use of constant expressions in conditions
  }
}
```
**What did you do? Please include the actual source code causing the issue.**
```js
if(false && abc==='str'){
}
```
**What did you expect to happen?**
Expect it to throw error
**What actually happened? Please include the actual, raw output from ESLint.**
no error

### PR DESCRIPTION
**Pull request by pvamshi**

### CODE REVIEW

**Comment by pvamshi** [comment]
Older pull request #5736 is closed when I closed the branch . Again sorry for new pull request, could not merge the changes gracefully .

File: lib/rules/no-constant-condition.js [hunk 1]
```diff
@@ -16,6 +16,27 @@ module.exports = function(context) {
...
+    /**
+     * Checks if a branch node of LogicalExpression short circuits the whole condition
+     * @param {ASTNode} node The branch of main condition which needs to be checked
+     * @param {string} operator The operator of the main LogicalExpression.
+     * @returns {boolean} true when condition short circuits whole condition
+     */
+    function isShortCircuit(node, operator) {
+        switch (node.type) {
+            case "Literal":
+                return (operator === "||" && node.raw === "true") ||
```
Comment (michaelficarra) [inline lib/rules/no-constant-condition.js]: Don't use `node.raw` for this.

Comment (pvamshi) [inline lib/rules/no-constant-condition.js]: Should it be `node.value`?

File: lib/rules/no-constant-condition.js [hunk 2]
```diff
@@ -16,6 +16,27 @@ module.exports = function(context) {
...
     //--------------------------------------------------------------------------
 
+
+    /**
+     * Checks if a branch node of LogicalExpression short circuits the whole condition
+     * @param {ASTNode} node The branch of main condition which needs to be checked
+     * @param {string} operator The operator of the main LogicalExpression.
+     * @returns {boolean} true when condition short circuits whole condition
+     */
+    function isShortCircuit(node, operator) {
```
Comment (michaelficarra) [inline lib/rules/no-constant-condition.js]: The term "short circuit" is not appropriate here. Short circuiting means the operator doesn't evaluate the right operand for some values of the left operand. This function is checking for the use of a logical identity.

File: lib/rules/no-constant-condition.js (hunk 1 above)
Comment (platinumazure) [inline lib/rules/no-constant-condition.js]: @pvamshi Yes, compare `node.value` to `true` (the boolean, not the string), that should work better.

Comment (michaelficarra) [inline lib/rules/no-constant-condition.js]: Yes. `node.value === true`.

File: lib/rules/no-constant-condition.js (hunk 2 above)
Comment (pvamshi) [inline lib/rules/no-constant-condition.js]: We are checking logical identity along with the operator . 
`|| + true` is true 
`&& + false` is true  , other combinations are not valid . So gave the shortCircuit name . Does `isLogicalIdentity` still valid ?

File: lib/rules/no-constant-condition.js (hunk 1 above)
Comment (pvamshi) [inline lib/rules/no-constant-condition.js]: Sure , added accordingly . Thanks for your guidance :)

File: lib/rules/no-constant-condition.js (hunk 2 above)
Comment (platinumazure) [inline lib/rules/no-constant-condition.js]: I think the problem is that `a && false` is a logical identity but not a short circuit. If you only check `false && a`, that's a short circuit (and a logical identity).

Comment (pvamshi) [inline lib/rules/no-constant-condition.js]: In current implementation `typeof abc==='str' || true` gives warning . Should we make sure this should not give error ?

**Comment by michaelficarra** [comment]
After re-reading the issue this PR claims to fix, I think this PR has missed the mark. What I believe @nzakas accepted was *only* reporting short-circuiting, not reporting when part of the condition is unnecessary, as he made clear in https://github.com/eslint/eslint/pull/5736#issuecomment-203634460. So this enhancement should be very simple: treat `false && X` and `true || X` as constant conditions for any `X`.

**Comment by pvamshi** [comment]
@michaelficarra 
If I understand correctly, 
This PR shows errors even for `X && false` and `X || true` also. I should avoid that , right ?

**Comment by michaelficarra** [comment]
Correct. It'd be nice to report those, but I agree with @nzakas that it should probably be a separate rule.

**Comment by pvamshi** [comment]
```js
var abc = 5;
if(abc ===5 && false){}
```
Still gives false. I am missing something . `false && X` and `X && false` both gives false right ? 

Similarly, `true || X` and `X || true` always gives true. So order should not matter right . 

The PR will not give error on `true && X` or `false || X` . Added test cases for them.

**Comment by michaelficarra** [comment]
I've just been basing this on @nzakas specifically calling out short circuiting in https://github.com/eslint/eslint/issues/5693#issuecomment-202566368. I think we should wait until he clarifies before making further changes.

**Comment by nzakas** [comment]
> Still gives false. I am missing something . false && X and X && false both gives false right ?

Yes, so those should both be flagged.

> Similarly, true || X and X || true always gives true. So order should not matter right .

Correct.

**Comment by pvamshi** [comment]
I think the PR is right . Please let me know if I am missing something :)

File: lib/rules/no-constant-condition.js (hunk 2 above)
Comment (nzakas) [inline lib/rules/no-constant-condition.js]: Agree with @michaelficarra, the name is a bit misleading. I think isLogicalIdentity is good.

Comment (pvamshi) [inline lib/rules/no-constant-condition.js]: Ok. Will change accordingly.

**Comment by nzakas** [comment]
Lgtm. Thanks!

**Pull request merged**